### PR TITLE
Improve Http2 ping timeout logic

### PR DESCRIPTION
**Pull request by akidambisrinivasan**
## Motivation and Context
Current PING timeout logic doesnt work well when there are delays in scheduling or flushing to socket.
```
Scenario we want to avoid is shown below (NOTE: ptm - pingTimeoutMillis, Wx - Write Ping x, Fx - Flush Ping x, Rx - Receive ack x, T -Timeout)

When timer is scheduled periodically, even though ack2 comes < ptm after being flushed to the socket, we
will still timeout at 2ptm.        
 0          1ptm       2ptm       3ptm
|----------|----------|----------|-------> time
W1F1 R1    W2      F2 T  R2
When timer is scheduled after flushing, we allow time for ack to come back and not prematurely timeout.
0            ptm1               ptm2
|-|----------|------|----------|-----------> time
 W1F1 R1      W2     F2   R2    W3
```
The change prevents premature closing of the HTTP2 connection. With this change, it will detect dead connections which are otherwise not detected due to inactivity while tolerating scheduling and flushing delays.

## Modifications
There are two modifications in this patch:
1. The callback to send the next ping is scheduled when the ping write is flushed, so it accounts for delays in flushing (because that delay doesnt mean the connection is dead)
2. And reset the ackReceivedTime for any data packet received not just ACK. That way if the channel has been seeing data, it is not really dead and it would be safe to wait for ACK for longer.

## Testing
Added 3 unit tests and modified one:
1. Add unit test to ensure that when there is a delay in writing to socket, if the ACK doesnt come within pingTimeoutMillis, the code still executes timeout logic and terminates the connection, but if it does come, resets the timer and continues pinging.
2. If there is a scheduling delay, it does not execute timeout. For scheduling delay, the system of jvm is unable to catchup and will do so slowly, the true issue that needs to be fixed is by addressing the slowness not timing out the connection.
3. When data packets are received, it represents activity which can be used to extend the timer for which the connection must be inactive.

## Types of changes

- [ x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by cenedhryn** [comment]
Thanks for providing the PR! We'll try to review it as soon as possible.

**Comment by cenedhryn** [comment]
The builds fail due to failed netty-nio-client tests:
~~~
[INFO]
--
450 | [INFO] Results:
451 | [INFO]
452 | [ERROR] Failures:
453 | [ERROR]   PingTimeoutTest.pingHealthCheck_10sec_shouldThrowExceptionAfter10Secs:118
454 | Expecting actual:
455 | 20.007813S
456 | to be between:
457 | [10S, 12S]
458 |  
459 | [ERROR]   PingTimeoutTest.pingHealthCheck_null_shouldThrowExceptionAfter5Sec:109
460 | Expecting actual:
461 | 10.006042S
462 | to be between:
463 | [5S, 7S]
464 |  
465 | [ERROR]   ServerNotRespondingTest.connectionNotAckPing_newRequestShouldUseNewConnection:121
466 | expected: 2
467 | but was: 1
468 | [ERROR]   Http2PingHandlerTest.nonAckPingsResultInOneChannelException:173
469 | Expected size: 1 but was: 0 in:
470 | []
471 | [INFO]
472 | [ERROR] Tests run: 331, Failures: 4, Errors: 0, Skipped: 1


~~~
I also checked out the code and ran it locally and I got 4 failures as well.

**Comment by debora-ito** [comment]
@akidambisrinivasan let us know if you want to keep this PR rolling.

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/FutureCancelHandlerTest.java [hunk 1]
```diff
@@ -78,7 +78,6 @@ public void methodSetup() {
         when(ctx.channel()).thenReturn(channel);
         when(channel.attr(EXECUTION_ID_KEY)).thenReturn(attrMap.attr(EXECUTION_ID_KEY));
         when(channel.attr(REQUEST_CONTEXT_KEY)).thenReturn(attrMap.attr(REQUEST_CONTEXT_KEY));
-        when(channel.id()).thenReturn(DefaultChannelId.newInstance());
```
Comment (cenedhryn) [inline http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/FutureCancelHandlerTest.java:81]: Why did you remove this line from the code?

Comment (akidambisrinivasan) [reply http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/FutureCancelHandlerTest.java:81]: Because I got this error

```
[ERROR] Errors:
[ERROR]   FutureCancelHandlerTest » UnnecessaryStubbing
Unnecessary stubbings detected in test class: FutureCancelHandlerTest
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at software.amazon.awssdk.http.nio.netty.internal.FutureCancelHandlerTest.methodSetup(FutureCancelHandlerTest.java:81)
Please remove unnecessary stubbings or use 'lenient' strictness. More info: javadoc for UnnecessaryStubbingException class.
[INFO]
[ERROR] Tests run: 401, Failures: 0, Errors: 1, Skipped: 1
```

File: core/sdk-core/src/test/java/software/amazon/awssdk/core/internal/http/timers/HttpClientApiCallTimeoutTest.java [hunk 2]
```diff
@@ -56,6 +56,8 @@ public class HttpClientApiCallTimeoutTest {
 
     private AmazonSyncHttpClient httpClient;
 
+    private final static int EPSILON_MILLIS = 10;
```
[the remark this answers is not in the document]
Comment (akidambisrinivasan) [reply core/sdk-core/src/test/java/software/amazon/awssdk/core/internal/http/timers/HttpClientApiCallTimeoutTest.java:59]: Because there will be slight variation in exactly when the scheduler runs.

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/fault/PingTimeoutTest.java [hunk 3]
```diff
@@ -99,23 +99,37 @@ public void methodTeardown() throws InterruptedException {
...
+     * should timeout during the second interval which is 10 seconds.
+     */
     @Test
     public void pingHealthCheck_10sec_shouldThrowExceptionAfter10Secs() {
         Instant a = Instant.now();
         assertThatThrownBy(() -> makeRequest(Duration.ofSeconds(10)).join()).hasCauseInstanceOf(IOException.class)
                                                                             .hasMessageContaining("An error occurred on the connection")
                                                                             .hasRootCauseInstanceOf(PingFailedException.class);
-        assertThat(Duration.between(a, Instant.now())).isBetween(Duration.ofSeconds(10), Duration.ofSeconds(12));
+        assertThat(Duration.between(a, Instant.now())).isBetween(Duration.ofSeconds(20), Duration.ofSeconds(22));
```
[the remark this answers is not in the document]
Comment (akidambisrinivasan) [reply http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/fault/PingTimeoutTest.java:132]: The default is 5 so I think testing is 5 is not improving coverage, I can make it 3 maybe

**Review by cenedhryn — APPROVED** [review]
